tower-http/src/classify.rs

```diff
@@ -0,0 +1,316 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Request, Response, StatusCode};
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for producing response classifiers from a request.
+///
+/// This is useful if your classifier depends on data from the request. Could for example be the
+/// URI or HTTP method.
+///
+/// This trait is generic over the error type your services produce. This is necessary for
+/// [`ClassifyResponse::classify_error`].
+pub trait MakeClassifier<E> {
+    /// The response classifier produced.
+    type Classifier: ClassifyResponse<
+        E,
+        FailureClass = Self::FailureClass,
+        ClassifyEos = Self::ClassifyEos,
+    >;
+
+    /// The type of failure classifications.
+    type FailureClass;
+
+    /// The type used to classify the response end of stream (EOS).
+    type ClassifyEos: ClassifyEos<E, FailureClass = Self::FailureClass>;
+
+    /// Make a response classifier for this request
+    fn make_classifier<B>(&self, req: &Request<B>) -> Self::Classifier;
```

Comment:
there should probably either be a blanket impl for `F: Fn(Request<B>) -> ClassifyResponse<E>`, or an adapter for making a function into a `MakeClassifier`?

can add that separately of course.

Reply:
Good point 👍 Made an issue for it https://github.com/tower-rs/tower-http/issues/74